lib/utils/ignore.ts

```diff
@@ -322,19 +291,44 @@ export async function filterFilesWithDockerignore(
 	const dockerignoreServiceDirs: string[] = multiDockerignore
 		? Object.keys(ignoreByDir).filter((dir) => dir && dir !== root)
 		: [];
-
-	const files = await listFiles(projectDir);
 	const dockerignoreFiles: FileStats[] = [];
-	const filteredFileList = files.filter((file: FileStats) => {
-		if (path.basename(file.relPath) === '.dockerignore') {
-			dockerignoreFiles.push(file);
-		}
-		for (const dir of dockerignoreServiceDirs) {
-			if (file.relPath.startsWith(dir)) {
-				return !ignoreByDir[dir].ignores(file.relPath.substring(dir.length));
-			}
-		}
-		return !ignoreByDir[root].ignores(file.relPath);
+	const filteredFileList: FileStats[] = [];
+	const klaw = await import('klaw');
```

Comment:
`klaw` was [already being used in the implementation of `originalTarDirectory`](https://github.com/balena-io/balena-cli/blob/v12.27.4/lib/utils/compose.js#L124), although not in a way that saves memory, as a full list of files was being computed before filtering, much like `listFiles()` did. However, `klaw` provides a stream interface to a non-recursive directory traversal that allows gradual filtering out of files as they are traversed. This is what I've done in this commit. It should mean a lower peak memory usage when traversing thousands of files that should be ignored (several `node_modules` folders).